Sources/Core/Microsoft.StreamProcessing/Ingress/SubscriptionBase.tt

```diff
@@ -273,10 +288,14 @@ foreach (bool disordered in new[] { true, false })
 
         /// <summary>
         /// Currently for internal use only - do not use directly.
+        /// Tracks the three possible low watermark values:
+        /// 1. rawValue - the raw low watermark value
+        /// 2. quantizedForLowWatermarkGeneration - raw low watermark value quantized for low watermark generation
+        /// 2. quanitzedForPunctuationGeneration - raw low watermark value quantized for punctuation generation
         /// </summary>
         [EditorBrowsable(EditorBrowsableState.Never)]
         [DataMember]
-        protected long lowWatermark = 0;
+        protected (long rawValue, long quantizedForLowWatermarkGeneration, long quantizedForPunctuationGeneration) lowWatermark = (0, 0, 0);
```

Comment:
I totally get why this is here, I just feel sad. It seems to be a mark of over-complexity.

Reply:
open to suggestions :)

---
In reply to: [320964086](https://github.com/microsoft/Trill/pull/108#discussion_r320964086) [](ancestors = 320964086)